rfd/0050-join-methods.md

```diff
@@ -0,0 +1,117 @@
+
+#### Proxy `POST /webapi/host/credentials`
+Join methods: token, EC2
+Server trust: PKI
+This is the proxy endpoint for registering IoT nodes that don't have a direct connection to the auth server, it basically forwards to
+the Auth `/tokens/register` endpoint.
+
+### IAM Join Method
+The IAM join method (introduced in RFD 41) requires gRPC methods rather than
+HTTP because the design requires streams to implement a challenge/response
+protocol. To summarize, after the client initiates the rpc the Auth server sends
+a crypto-random "challenge" string over the server->client stream. This
+challenge must be included in a signed `sts:GetCallerIdentity` request which
+will be sent on the client->server stream.  Finally, if everything checks out,
+the Auth server will send the signed certificates on the server->client stream.
+
+A new gRPC streaming method will be added on Auth and Proxy to support this. The
+client will be able to call the gRPC method with either a Proxy or Auth address
+and it will "just work" transparently.
+
+#### Auth `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
+Join mathods: IAM
+Server trust: CA pins
+This is the Auth endpoint that will complete the IAM join request.
+
+#### Proxy `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
+Join mathods: IAM
+Server trust: PKI
+Normally, authenticated gRPC calls from IoT nodes are tunnelled through the
+proxy over SSH. This will not work for unauthenticated clients which don't yet
+have an SSH certificate.
+
+It would be possible to use TLS routing on the proxy to forward gRPC requests
+directly to the Auth server, but this would open up the Auth server to DOS
+attacks from unauthenticated clients.
+
+Instead, this single gRPC method will be implemented on the proxy. This method
+will first perform rate-limiting, then "forward" the request to the auth server
+method at the application layer by calling the gRPC method through the Proxy's
+own Auth client.
+
+Since the Proxy currently does not expose any gRPC service of its own, a new
+gRPC listener will be added to support this. gRPC connections will be
+multiplexed on the web listener/port using the existing ALPN-based multiplexer
+and a special ALPN ProtocolName of `teleport-proxy-grpc` that will be passed by
+the client.
+
+### Certificate Bot
+The Certificate Bot needs to get an initial certificate for the cluster which
+will be renawable. It will either provide a token or use the new EC2 or IAM
+methods to get the initial certificate. This is very similar to a node joining a
+cluster, except that the obtained certificate will be renewable.
+
+We can re-use the existing `/tokens/register` (on Auth) and
+`/webapi/host/credentials` (on Proxy) endpoints and extend the backend
+implementations to provide renewable certs if allowed by the corresponding token
+in the backend. This will also allow the client-side logic on the bot to reuse
```

Comment:
I'm curious how we'll want to differentiate backend implementations. These endpoints today assume a `ProvisionToken` where the bot currently uses a `UserToken` (with a sub-kind of `bot`).

We'll definitely want to reuse `ProvisionTokenV2` for EC2/IAM joining so maybe it'd be best to adopt them for regular token joining too? It's arbitrary from the bot's end but we do need to know that the token is specifically a bot token, and for which user it's intended. Could we reuse the `bot_user` field from below for this?

Reply:
Yeah from our brief discussion I thought the idea was to use `ProvisionTokenV2` for bots. If we're going to do it for EC2/IAM methods, it would make sense to do it for the secret token method as well IMO.

Although, the `.spec.roles` field currently must be a `SystemRole` (`proxy`, `auth`, etc.) so I'm not sure what to do about this for the bot. There is already `RoleSignup` for when user tokens are converted to machine tokens. Maybe we could make `RoleBot` for and use that to differentiate bot tokens within `ProvisionTokenV2`.

Reply:
I think this is reasonable assuming we can add a new `SystemRole` without much hassle.

It might be a bit of a hack, but as an alternative we could pick our implementation based on whether the `bot_user` field is set.